lib/libc/gen/getusershell.3
.\"	BSDI getusershell.3,v 2.1 1995/02/03 06:22:29 polk Exp
.\"
.\" Copyright (c) 1985, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)getusershell.3	8.1 (Berkeley) 6/4/93
.\"
.Dd June 4, 1993
.Dt GETUSERSHELL 3
.Os BSD 4.3
.Sh NAME
.Nm getusershell ,
.Nm setusershell ,
.Nm endusershell
.Nd get legal user shells
.Sh SYNOPSIS
.Fd #include <unistd.h>
.Ft char *
.Fn getusershell void
.Ft void
.Fn setusershell void
.Ft void
.Fn endusershell void
.Sh DESCRIPTION
The
.Fn getusershell
function
returns a pointer to a legal user shell as defined by the
system manager in the file 
.Pa /etc/shells .
If 
.Pa /etc/shells
is unreadable or does not exist,
.Fn getusershell
behaves as if
.Pa /bin/sh
and
.Pa /bin/csh
were listed in the file.
.Pp
The
.Fn getusershell
function
reads the next
line (opening the file if necessary);
.Fn setusershell
rewinds the file;
.Fn endusershell
closes it.
.Sh FILES
.Bl -tag -width /etc/shells -compact
.It Pa /etc/shells
.El
.Sh DIAGNOSTICS
The routine
.Fn getusershell
returns a null pointer (0) on
.Dv EOF .
.Sh SEE ALSO
.Xr shells 5
.Sh HISTORY
The
.Fn getusershell
function appeared in 
.Bx 4.3 .
.Sh BUGS
The
.Fn getusershell
function leaves its result in an internal static object and returns
a pointer to that object. Subsequent calls to
.Fn getusershell
will modify the same object.


libexec/ftpd/support/grot/getusershell.c
/*
 * Copyright (c) 1985 Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms are permitted
 * provided that: (1) source distributions retain this entire copyright
 * notice and comment, and (2) distributions including binaries display
 * the following acknowledgement:  ``This product includes software
 * developed by the University of California, Berkeley and its contributors''
 * in the documentation or other materials provided with the distribution
 * and in all advertising materials mentioning features or use of this
 * software. Neither the name of the University nor the names of its
 * contributors may be used to endorse or promote products derived
 * from this software without specific prior written permission.
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 */

#if defined(LIBC_SCCS) && !defined(lint)
static char sccsid[] = "@(#)getusershell.c	5.6 (Berkeley) 6/1/90";

#endif /* LIBC_SCCS and not lint */

#include "../src/config.h"
#include <sys/types.h>
#include <sys/param.h>
#include <sys/file.h>
#include <sys/stat.h>
#include <stdlib.h>
#include <ctype.h>
#include <stdio.h>

#define SHELLS "/etc/shells"

/*
 * Do not add local shells here.  They should be added in /etc/shells
 */
static char *okshells[] =
{"/bin/sh", "/bin/csh", 0};

static char **shells,
 *strings;
static char **curshell = NULL;
static char **initshells();

/*
 * Get a list of shells from SHELLS, if it exists.
 */
char *
getusershell(void)
{
	char *ret;

	if (curshell == NULL)
		curshell = initshells();
	ret = *curshell;
	if (ret != NULL)
		curshell++;
	return (ret);
}

void
endusershell(void)
{
	if (shells != NULL)
		free((char *) shells);
	shells = NULL;
	if (strings != NULL)
		free(strings);
	strings = NULL;
	curshell = NULL;
}

void
setusershell(void)
{
	curshell = initshells();
}

static char **
initshells(void)
{
	register char **sp,
	 *cp;
	register FILE *fp;
	struct stat statb;

	if (shells != NULL)
		free((char *) shells);
	shells = NULL;
	if (strings != NULL)
		free(strings);
	strings = NULL;
	if ((fp = fopen(SHELLS, "r")) == (FILE *) 0)
		return (okshells);
	if (fstat(fileno(fp), &statb) == -1) {
		(void) fclose(fp);
		return (okshells);
	}
	if ((strings = (char *) malloc((unsigned) statb.st_size + 1)) == NULL) {
		(void) fclose(fp);
		return (okshells);
	}
	shells = (char **) calloc((unsigned) statb.st_size / 3, sizeof(char *));

	if (shells == NULL) {
		(void) fclose(fp);
		free(strings);
		strings = NULL;
		return (okshells);
	}
	sp = shells;
	cp = strings;
	while (fgets(cp, MAXPATHLEN + 1, fp) != NULL) {
		while (*cp != '#' && *cp != '/' && *cp != '\0')
			cp++;
		if (*cp == '#' || *cp == '\0')
			continue;
		*sp++ = cp;
		while (!isspace(*cp) && *cp != '#' && *cp != '\0')
			cp++;
		*cp++ = '\0';
	}
	*sp = (char *) 0;
	(void) fclose(fp);
	return (shells);
}
